adding-support-for-alternative-ais

RayOffiah opened the pull request:
Making the code a bit more modular so that we can add support for other AI models.

RayOffiah committed:
Updated the query for Couchbase Operator.

RayOffiah committed:
Update AI configuration handling for release notes generation

Refactored AI-related settings to support multiple services and models, replacing `ai_prompt` and `openai_api_key` with a more flexible `ai_service` object. Updated configuration schema, YAML files, and logic to integrate these changes. Added reusable methods for password handling and improved the overall structure for clarity.

RayOffiah committed:
Refactor AI client architecture for better modularity

Introduced an `AIObject` module with an abstract base class and concrete implementation for different AI services. Replaced direct OpenAI client calls with a factory method to support extensibility. Adjusted related YAML configuration and updated `cb_release_note.py` to utilize the new architecture.

osfameron commented on requirements.txt:
```diff
@@ -9,4 +9,5 @@ termcolor~=2.1.1
 alive-progress~=3.2.0
 openai~=1.70.0
 pyinstaller~=6.11.1
+copilot-api~=1.0.0
```
doesn't look like this module is used yet, is that for future expansion?
(Also, how do you get access to copilot API? It doesn't seem to be available at all via Github)

RayOffiah replied:
Yeah, that's for the future. 

I don't have a key for that yet. 

I just wanted to get the code ready while I was having a good day.  🙂

osfameron commented on AIObject.py:
```diff
@@ -0,0 +1,41 @@
+
+
+
+class OpenAIClient(AbstractAIO):
```
Could this not just be a function that takes the `ai_service` from the yaml as a dictionary input? 

(Not a problem if the Enterprise OO constructs like abstract base classes and factories are earning their keep, or you think will be soon, but ideally all the tooling should be as simple to understand and modify as possible)

RayOffiah replied:
Mmm. That's a good point. 🤔

I could just pass in the whole `ai_service`. 

I'll fix it.

RayOffiah replied:
I'm going to keep the structure as is for the time being at least, until I've had a chance to play with other AI setups. This does keep a lot of the complexity in the factory. (That's the plan anyway)

RayOffiah committed:
Refactor AI client handling and improve naming consistency

Renamed `AIObject` module to `AI_Clients` and revised class names for clarity and consistency. Updated `ai_client_factory` to simplify parameter usage and enhance modularity. Adjusted `cb_release_note.py` to align with the improved architecture and naming changes.

RayOffiah committed:
Refactor template handling, AI client architecture, and JQL queries

Simplified Jinja2 template logic and introduced a new `couchbase-docs-server-whats-new` template for improved maintainability. Streamlined AI client implementation by removing the abstract base class and optimizing the `OpenAIClient`. Updated `.gitignore` rules for better adoc file pattern handling. Enhanced JQL queries with additional conditions for Couchbase Server and Operator.

RayOffiah committed:
Refactor configuration management, templates, and AI client integration

Replaced OpenAI with modular AI client architecture supporting multiple services, including Google Gemini. Updated YAML configurations for enhanced clarity and JQL query accuracy. Simplified and reordered templates for `couchbase-mobile` and `docs-server` while improving naming consistency. Updated packaging with new dependencies and refined method signatures for better type hinting and maintainability.

RayOffiah committed:
Add return type annotations to `get_ai_response` method across AI clients

Refactored `get_ai_response` in `AIClient`, `GeminiClient`, and `OpenAIClient` to include explicit return type hints for improved code clarity and maintainability.

RayOffiah committed:
Update AI service in `cb_release_notes_config.yaml` to use ChatGPT with GPT-4o and refine asciidoc prompt style

RayOffiah committed:
Remove Backup Service issue table from `couchbase-docs-server.jinja2` and refine asciidoc prompt style in `cb_release_notes_config.yaml`

RayOffiah committed:
Add `couchbase-docs-server-old` template and expand `Tools` issue table scope in `couchbase-docs-server`

RayOffiah committed:
Remove `couchbase-docs-server-old` template and narrow `Tools` issue table scope in `couchbase-docs-server`

RayOffiah committed:
Expand `Views` issue table scope with the `UI` tag, enhance AI summary generation with service metadata, and refine AI prompt style in YAML configuration.
